public/app/features/scenes/components/layout/SceneGridLayout.tsx

```diff
@@ -0,0 +1,574 @@
+        child.setState({
+          size: {
+            ...child.state.size,
+            x: childLayout.x,
+            y: row ? childLayout.y - row.state.size!.y! - 1 : childLayout.y,
+          },
+        });
+      }
+    }
+
+    // Skipping onLayoutChange handler to avoid unnecessary re-renders
+    this._skipOnLayoutChange = true;
+    this.updateLayout();
+  };
+}
+
+function SceneGridLayoutRenderer({ model }: SceneComponentProps<SceneGridLayout>) {
+  const theme = useTheme2();
+  const { children } = model.useState();
+  validateChildrenSize(children);
+
+  const layout = useMemo(() => {
+    let topLevelCells: ReactGridLayout.Layout[] = [];
+    const perRowCells: Record<string, ReactGridLayout.Layout[]> = {};
+
+    Object.values(model.flattenedChildren).forEach((child) => {
+      const size = child.child.state.size!;
+
+      if (!child.row) {
+        // rendering cells and row
+        let isDraggable = Boolean(child.child.state.isDraggable);
+        let isResizable = Boolean(child.child.state.isResizable);
+        if (child.child instanceof SceneGridRow) {
+          isDraggable = child.child.state.isCollapsed ? true : false;
+          isResizable = false;
+        }
+        topLevelCells.push({
+          i: child.child.state.key!,
+          x: size.x!,
+          y: size.y!,
+          w: size.width!,
+          h: size.height!,
+          isResizable,
+          isDraggable,
+        });
+      } else {
+        // Rendering row children
+        let isDraggable = Boolean(child.child.state.isDraggable);
+        let isResizable = Boolean(child.child.state.isResizable);
+        const cell = {
+          i: child.child.state.key!,
+          x: size.x!,
+          // y: size.y!,
+          y: size.y! + child.row.state.size!.y!,
+          w: size.width!,
+          h: size.height!,
+          isResizable,
+          isDraggable,
+        };
+        if (perRowCells[child.row.state.key!]) {
+          perRowCells[child.row.state.key!].push(cell);
+        } else {
+          perRowCells[child.row.state.key!] = [cell];
+        }
+      }
+    });
+
+    topLevelCells = topLevelCells.sort((panelA, panelB) => {
+      if (panelA.y === panelB.y) {
+        return panelA.x - panelB.x;
+      } else {
+        return panelA.y - panelB.y;
+      }
+    });
+    // sort sub row cells
+    for (const rowKey in perRowCells) {
+      perRowCells[rowKey] = perRowCells[rowKey].sort((panelA, panelB) => {
+        if (panelA.y === panelB.y) {
+          return panelA.x - panelB.x;
+        } else {
+          return panelA.y - panelB.y;
+        }
+      });
+    }
+
+    const lg = topLevelCells.reduce<ReactGridLayout.Layout[]>((acc, cell) => {
+      if (perRowCells[cell.i]) {
+        return [...acc, cell, ...perRowCells[cell.i]];
+      }
+      return [...acc, cell];
+    }, []);
+
+    return { lg, sm: lg.map((l) => ({ ...l, w: 24 })) };
+  }, [model.flattenedChildren]);
```

Comment:
moving this this to the SceneObject class would make it easier to unit test. 

I can't say I understand the logic here, especially the topLevelCells, sub-row cells, the special sorting, the reduce.

Reply:
@torkelo  - I think this sort needs to happen in the flattening phase - as based on the flattened children the grid is rendered (to solve the same issue as with the current grid - to have DOM elements in the correct order for mobile view I suppose). 

As of the sort: 
1. `topLevelCells` contains all cells that are either row OR not within a row.
2. perRowCells - Record of each row cells.

So what happens here is:
1. sort the non-sub-row cells first 
2. sort cells PER row
3. Combine everything into sorted layout array.

This is problematic tho when the "responsive" view kicks in. as the DOM order and the layout order are different. Hence I think the flattened children need to be sorted already.

Reply:
ok, think I understand. But why do you need the perRowCells ?